File: src/main/java/com/morph/engine/util/Feed.java
package com.morph.engine.util;

import io.reactivex.Emitter;

import java.util.ArrayList;
import java.util.List;

public class Feed<T> {
    private List<Listener<T>> listeners = new ArrayList<>();

    public void register(Listener<T> listener) {
        listeners.add(listener);
    }

    public void onNext(T t) {
        listeners.forEach(listener -> listener.onNext(t));
    }

    public void onError(Throwable t) {
        listeners.forEach(listener -> listener.onError(t));
    }

    public void onComplete() {
        listeners.forEach(Listener::onComplete);
    }

    public void emit(Emitter<T> emitter) {
        Listener<T> listener = new Listener<T>() {
            @Override
            public void onNext(T t) {
                emitter.onNext(t);
            }

            @Override
            public void onError(Throwable t) {
                emitter.onError(t);
            }

            @Override
            public void onComplete() {
                emitter.onComplete();
            }
        };
        register(listener);
    }
}

File: src/main/kotlin/com/morph/engine/util/Feed.kt
package com.morph.engine.util

import io.reactivex.Emitter

class Feed<T> {
    private val listeners = ArrayList<Listener<T>>()

    fun register(listener: Listener<T>) {
        listeners.add(listener)
    }

    fun onNext(t: T) {
        listeners.forEach { listener -> listener.onNext(t) }
    }

    fun onError(t: Throwable) {
        listeners.forEach { listener -> listener.onError(t) }
    }

    fun onComplete() {
        listeners.forEach { it.onComplete() }
    }

    fun emit(emitter: Emitter<T>) {
        val listener = object : Listener<T> {
            override fun onNext(t: T) {
                emitter.onNext(t)
            }

            override fun onError(t: Throwable) {
                emitter.onError(t)
            }

            override fun onComplete() {
                emitter.onComplete()
            }
        }
        register(listener)
    }

    inline fun <reified R> emit(emitter: Emitter<R>, crossinline action: Emitter<R>.(T) -> Unit) {
        val listener = object : Listener<T> {
            override fun onNext(t: T) {
                emitter.action(t)
            }

            override fun onError(t: Throwable) {
                emitter.onError(t)
            }

            override fun onComplete() {
                emitter.onComplete()
            }
        }
        register(listener)
    }
}
